Bintools Wrapper: Init by refactoring out of cc-wrapper, take 2

Ericson2314 opened the pull request:
**edit** see my question in https://github.com/NixOS/nixpkgs/pull/29396#issuecomment-333688023

###### Motivation for this change

This is meant to be reviewed commit by commit.

This is needed so packages that just use binutils get the same setup-hook support. That is, variables like `BUILD_AS` or `LD` should be defined the same way, along with the `LD_FLAGS`.  This will be necessary to clean up at least the gcc derivation (which must be built with an binutils not associated with any existsing C compiler when target != host), and probably other compilers too.

More broadly, having two wrappers will help enforce proper layering between binutils and the C compiler. For languages besides C/C++, it would be neat to use `stdenvNoCC + binutils, avoiding a C compiler dependency. This wrapper makes that change possible.

@edolstra I'm hoping for you to merge this change when you approve of it, to avoid a repeat of last time.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built stdenv on platform(s)
   - [ ] NixOS --- in progress
   - [x] macOS
   - [ ] Linux 
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Ericson2314 committed:
bintools-wrapper: Import separately from cc-wrapper

Ericson2314 committed:
bintools-wrapper: Init

Factor a bintools (i.e. binutils / cctools) wrapper out of cc-wrapper. While
only LD is wrapped, the setup hook defines environment variables on behalf of
other utilites.

Ericson2314 committed:
cc-wrapper: Pull variable mangler into utils.sh

In preparation for splitting out bintools-wrapper

Ericson2314 committed:
cc-wrapper: Use separate mangler for "bool" variables

This avoids any `NIX_FOOBAR=1 1` not triggering conditions.

Ericson2314 committed:
doc: Document Bintools Wrapper

Shrunk the CC Wrapper documentation so as not to be repetative.

Ericson2314 commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -236,16 +196,25 @@ stdenv.mkDerivation {
+      # TODO(@Ericson2314): Remove this after stable release and force
+      # everyone to refer to binutils-wrapper directly.
+      if [[ -f "$binutils/nix-support/dynamic-linker" ]]; then
+        ln -s "$binutils/nix-support/dynamic-linker" "$out/nix-support"
```
@edolstra this symlink avoids the incompatibility the old version introduced.

orivej commented on pkgs/build-support/cc-wrapper/ld-wrapper.sh:
```diff
@@ -63,6 +63,21 @@ fi
+# Choose 32-bit dynamic linker if needed
+if [ -e @out@/nix-support/dynamic-linker-m32 ]; then
+    prev=
+    for p in ${params+"${params[@]}"}; do
```
It was determined to be valuable not to iterate over params more times than necessary. Could you merge this loop into the next one? (It has already merged two cases, `[ "$NIX_@infixSalt@_DONT_SET_RPATH" != 1 ]` and `[ "$NIX_@infixSalt@_SET_BUILD_ID" = 1 ]`.)

Ericson2314 replied:
Good idea!

Ericson2314 replied:
Done

orivej commented on pkgs/build-support/binutils-wrapper/ld-wrapper.sh:
```diff
@@ -67,8 +67,16 @@ declare -a libDirs
-if [ "$NIX_@infixSalt@_DONT_SET_RPATH" != 1 ] || [ "$NIX_@infixSalt@_SET_BUILD_ID" = 1 ]; then
+# Two tasks:
+#
+#   1. Find all -L... switches for rpath, and relocatable flags for build id.
```
These cases are not related, so if you enumerate them, they should be 1 and 2.

Ericson2314 replied:
True. Good catch.

Ericson2314 replied:
Done

orivej commented on pkgs/build-support/binutils-wrapper/setup-hook.sh:
```diff
@@ -0,0 +1,63 @@
+    case $depOffset in
+        -1) local role='BUILD_' ;;
+        0)  local role='' ;;
+        1)  local role='TARGET_' ;;
```
Where may `depOffset` be set to 1? I find only places where it is set to -1 or 0.

Ericson2314 replied:
Oh heh. That would anticipate https://github.com/NixOS/nixpkgs/pull/26805

Ericson2314 replied:
Cc-wrapper already has that, I guess I'll leave it as consistent anticipation.

DavidEGrayson commented on doc/stdenv.xml:
```diff
@@ -1343,37 +1342,62 @@ someVar=$(stripHash $name)
-        CC wrapper's setup hook causes any <filename>include</filename> subdirectory of such a dependency to be added to <envar>NIX_CFLAGS_COMPILE</envar>, and any <filename>lib</filename> and <filename>lib64</filename> subdirectories to <envar>NIX_LDFLAGS</envar>.
-        The setup hook itself contains some lengthy comments describing the exact convoluted mechanism by which this is accomplished.
+        Binutils Wrapper's setup hook causes any <filename>lib</filename> and <filename>lib64</filename> subdirectories to <envar>NIX_LDFLAGS</envar>.
+        Sine CC Wrapper and Binutils Wrapper use the same strategy, most of the Binutils Wrapper code is sparsely commented and refers to CC Wrapper.
```
"Sine" is a typo.

Ericson2314 replied:
Thanks!

DavidEGrayson commented on doc/stdenv.xml:
```diff
@@ -1343,37 +1342,62 @@ someVar=$(stripHash $name)
         It is currently accomplished by collecting directories of host-platform dependencies (i.e. <varname>buildInputs</varname> and <varname>nativeBuildInputs</varname>) in environment variables.
-        CC wrapper's setup hook causes any <filename>include</filename> subdirectory of such a dependency to be added to <envar>NIX_CFLAGS_COMPILE</envar>, and any <filename>lib</filename> and <filename>lib64</filename> subdirectories to <envar>NIX_LDFLAGS</envar>.
-        The setup hook itself contains some lengthy comments describing the exact convoluted mechanism by which this is accomplished.
+        Binutils Wrapper's setup hook causes any <filename>lib</filename> and <filename>lib64</filename> subdirectories to <envar>NIX_LDFLAGS</envar>.
```
Probably you're missing the words "be added to" in this sentence.

DavidEGrayson commented on doc/stdenv.xml:
```diff
@@ -1343,37 +1342,62 @@ someVar=$(stripHash $name)
         Firstly, this helps poorly-written packages, e.g. ones that look for just <command>gcc</command> when <envar>CC</envar> isn't defined yet <command>clang</command> is to be used.
-        Secondly, this helps packages not get confused when cross-compiling, in which case multiple CC wrappers may be simultaneous in use (targeting different platforms).
-        <envar>BUILD_</envar>- and <envar>TARGET_</envar>-prefixed versions of the normal environment variable are defined for the additional CC Wrappers, properly disambiguating them.
+        Secondly, this helps packages not get confused when cross-compiling, in which case multiple Binutils Wrappers may be simultaneous in use (targeting different platforms).
```
The idomatic wording options would be:

- "may be in use simultaneously", or
- "may be in simultaneous use"

DavidEGrayson commented on doc/stdenv.xml:
```diff
@@ -1343,37 +1342,62 @@ someVar=$(stripHash $name)
+      </para>
+      <para>
+        Dependency finding is undoubtedly the main task of CC Wrapper.
+        This works just like Binutils Wrapper, except the any <filename>include</filename> subdirectory of any relevant dependency is added to <envar>NIX_CFLAGS_COMPILE</envar>.
```
"the" should be "that"

DavidEGrayson commented on doc/stdenv.xml:
```diff
@@ -1343,37 +1342,62 @@ someVar=$(stripHash $name)
+        A problem with this final task is that Binutils Wrapper is honest and defines <envar>LD</envar> as <command>ld</command>.
         Most packages, however, firstly use the C compiler for linking, secondly use <envar>LD</envar> anyways, defining it as the C compiler, and thirdly, only so define <envar>LD</envar> when it is undefined as a fallback.
-        This triple-threat means CC Wrapper will break those packages, as LD is already defined as the actually linker which the package won't override yet doesn't want to use.
+        This triple-threat means Binutils Wrapper will break those packages, as LD is already defined as the actually linker which the package won't override yet doesn't want to use.
```
"actually" is not an adjective, probably you meant "actual".

Ericson2314 committed:
misc setup-hooks: Use env vars to refer to binutils programs

This is more robust for cross-compilation

Ericson2314 commented:
[Blocked by https://github.com/NixOS/nixpkgs/pull/29568]

Ericson2314 commented:
I'm rebuilding now, but I've built other PRs containing this so quite confident it will succeed. This is ready for final review!

Ericson2314 commented on pkgs/build-support/binutils-wrapper/setup-hook.sh:
```diff
@@ -0,0 +1,63 @@
+done
+
+# No local scope in sourced file
+unset role cmd
```
@edolstra added a commit to fix the `CMD=windres` issue you pointed out in https://github.com/NixOS/nixpkgs/issues/29846 .

Ericson2314 commented:
On second thought, I'd like to rename this `utils-wrapper`, and `stdenv.cc.utils`. "cctools + GNU binutils" is also wrapped, and isn't "GNU Binutils" so I think a name like "utils" would be better.

copumpkin commented:
@Ericson2314 that seems fine to me

Ericson2314 commented:
I'm going to wait to kick of a nixborg mass rebuild until https://github.com/NixOS/nixpkgs/pull/31775 is finished, but this is now again ready for rebuild.

Ericson2314 committed:
cc-wrapper: Don't treat "-" alone as a flag

It means stdin, and is morally equivalent to passing a file. e.g.

  $ echo 'int main(void) { return 0; }' | gcc -x c -

will compile and link a binary.

Ericson2314 commented:
@nixborg buidl

Ericson2314 commented:
@nixborg build

nixborg commented:
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-29396

nixborg commented:
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-29396

Ericson2314 commented:
@nixborg build

nixborg commented:
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-29396

Ericson2314 commented:
@nixborg build

nixborg commented:
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-29396

Ericson2314 commented on pkgs/build-support/cc-wrapper/add-flags.sh:
```diff
@@ -66,17 +55,9 @@ if [ -e @out@/nix-support/gnat-cflags ]; then
-fi
-
 if [ -e @out@/nix-support/cc-ldflags ]; then
     NIX_@infixSalt@_LDFLAGS+=" $(< @out@/nix-support/cc-ldflags)"
```
I tried to make this use `CFLAGS_LINK` instead of `LDFLAGS` for better layering, but that changed the priority breaking stdenv (as the bootstrap tools lib directory will resolve so much).

Ericson2314 commented:
I built Atlas locally (without any rebase, unlike nixborg) and it succeeded. This reassures me the build failure is spurious / from something else, not a problem with the wrapper and `gfortran` as I feared.

Ericson2314 committed:
bintools-wrapper: Support ld.ldd, along with ld.bfd and ld.gold

Also make the code more precise in the process

Ericson2314 committed:
darwin binutils: Better handling of man pages and info

Ericson2314 commented:
I've add a few more cleanups on top, doing another rebuild.

Ericson2314 commented:
@nixborg build

Ericson2314 commented:
Curious whether the last `LDEMULATION`-defining commit unbroke cross, so repurposed https://hydra.nixos.org/jobset/nixpkgs/ericson2314-cross-trunk . If not, I'll just give up on `--enable-targets=all` in binutils: it's less a panacea than I thought because the GNU Assembler doesn't become multi-target.

nixborg commented:
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-29396

Ericson2314 commented on pkgs/os-specific/darwin/binutils/default.nix:
```diff
@@ -37,10 +38,16 @@ stdenv.mkDerivation {
 
     ln -s ${cctools}/libexec $out/libexec
+
+    mkdir -p "info/nix-support" "man/nix-support"
```
@vcunat there's no nix-level way to do per-output propgated deps, right?

vcunat replied:
@Ericson2314: right, I think there's no better way than this ATM.  There certainly wasn't when multi-output stdenv was done (years ago), and I haven't seen any use case up to now.  If more of them appear, we can design some kind of API.

orivej commented:
> If not, I'll just give up on `--enable-targets=all` in binutils

I think multitarget binutils (enabled in https://github.com/NixOS/nixpkgs/pull/30484/commits/f8741c38cd546e3ff18ce9d708de14ff2aae68ab) is a good thing even if they can not be used for cross.

Ericson2314 commented:
@nixborg build

nixborg commented:
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-29396

Ericson2314 commented:
@nixborg build

nixborg commented:
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-29396

Ericson2314 commented:
@nixborg build

nixborg commented:
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-29396

Ericson2314 committed:
bintools-wrapper: Create man and info outputs propagated underlying ones

These will be installed if the wrappers are. The wrappers aren't very
good to install, but that's another matter.

Ericson2314 commented:
https://github.com/NixOS/nixpkgs/pull/26805 is needed to fix progated inputs, so there's little point trying cross with this PR alone. I moved the `LDEMULATION` commit to that other PR accordingly. I also pullled  some non-controversial stuff out of the end of this into https://github.com/NixOS/nixpkgs/issues/28448 into https://github.com/NixOS/nixpkgs/issues/28448.

I'm looking to merge this within the week, so I can get onward to the more interesting cross stuff. I now count 3 separate instances of people decently blocked on me unbreaking things, and this has sat open for a while. @edolstra you're the point person for a change this big, so let me know what I can do to help you review.

vcunat reviewed (approved):
OK, I think.

I read though this all, not finding any problem (beyond the nitpicks I commented on); but I can't say I fully understand _all_ the necessary bits.

orivej commented:
These breaks the build of `libstdcxx5`.

Ericson2314 commented on pkgs/build-support/cc-wrapper/utils.sh:
```diff
@@ -1,20 +1,35 @@
+    local -a role_infixes=("$@")
+
+    local outputVar="${var/+/_@infixSalt@_}"
+    declare -gxi ${outputVar}+=''
```
That's an oversight, will fix.

Ericson2314 commented on pkgs/build-support/bintools-wrapper/default.nix:
```diff
@@ -82,7 +82,8 @@ stdenv.mkDerivation {
 
-  outputs = [ "out" "man" ];
+  outputs  = [ "out" "info" "man" ];
+  outputsToInstall = [ "out" "info" "man" ];
```
What do you mean by `programs.man.enable`? I'll gladly just remove `outputsToInstall`, or do whatever you think is best, just confused what that is.

Ericson2314 commented on doc/stdenv.xml:
```diff
@@ -1351,36 +1351,64 @@ someVar=$(stripHash $name)
-        Secondly, this helps packages not get confused when cross-compiling, in which case multiple CC wrappers may be simultaneous in use (targeting different platforms).
-        <envar>BUILD_</envar>- and <envar>TARGET_</envar>-prefixed versions of the normal environment variable are defined for the additional CC Wrappers, properly disambiguating them.
+        Secondly, this helps packages not get confused when cross-compiling, in which case multiple Bintools Wrappers may simultaneously be in use.
+        <footnote><para>Each wrapper targets a single platform, so if binaries for multiple platforms are needed, multiple wrappers are needed must be used.</para></footnote>
```
Thanks!

Ericson2314 committed:
gcc: Adjust builder.sh to find some things in bintools-wrapper instead

Ericson2314 commented:
@nixborg build

nixborg commented:
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-29396

Ericson2314 commented:
@grahamc what is the `nix-instantiate ...` that's failing? the ellipses won't expand.

grahamc commented:
If you triple-click the text it'll select it all and you can copy it out: `grahamcofborg-eval-nixpkgs-unstable-jobset — nix-instantiate ./pkgs/top-level/release.nix -A unstable `

Ericson2314 commented:
@grahamc thanks!

Ericson2314 committed:
cc-wrapper: GNAT wrapper stop caring about `-m32`

It need not concern itself with 32-bit dynamic linking; ld-wrapper handles that now.

Ericson2314 commented:
Ah, I did `dynamic-linker-32` not `dynamic-linker-m32` in the fix. @orivej I think I fixed those issues you mentioned.

Ericson2314 commented:
@nixborg build

nixborg commented:
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-29396

Ericson2314 committed:
libstdc++5: Fix after binutils-wrapper

Will need to be editted again to work for cross

Ericson2314 commented:
@orivej both are now fixed

Ericson2314 committed:
clang multi: Fix post bintools wrapper

Ericson2314 committed:
Merge commit '9d8f9b2e531bf95a700a949d879927fb6996ffc9' into binutils-wrapper

Ericson2314 committed:
Merge remote-tracking branch 'upstream/staging' into binutils-wrapper

Ericson2314 commented:
@edolstra told me he doesn't have time to review this any time soon, so we might as well not hold it up further. Yay, thanks!

This has passed nixborg CI many times before. the last change is a mass rebuild, but just from merging with staging. The post-merge change is just an eval fix for a newer package (clang multilib). Merging, but will keep an I on staging just in case.

orivej commented:
This bloke `bloaty`, fixed in fe61c3b84e8e81a8ec2bf6b3ed2a0e8652cea190.

orivej commented:
This broke `lispPackages.cl-fuse`. @Ericson2314 could you fix it? It exports `NIX_LDFLAGS` with `-L/nix/store/wr579906xik1qbnblm0wq1nissxh53sy-fuse-2.9.7/lib` and runs `gcc` (wrapper), but `gcc` invokes unwrapped `gcc` without these `NIX_LDFLAGS`. Here is this sequence of steps: https://gist.github.com/orivej/4c8fc1b12d8d630edf24b6ff5f88cd80 (`141-142-sbcl` runs `sbcl`, which runs `gcc` in the same way as `142-143-gcc`, which runs the unwrapped `gcc` just like `143-144-gcc `.) Here is how `NIX_LDFLAGS` are defined: https://github.com/NixOS/nixpkgs/blob/ea15e43693908291b48f3852c32721a085aa8f84/pkgs/development/lisp-modules/define-package.nix#L23

orivej commented:
Two perl packages,`perlPackages.EncodeDetect` and `perlPackages.ExtUtilsCppGuess`, fail to build with `ld: cannot find -lstdc++`. They are needed for `spamassassin` and `slic3r`. @Ericson2314 agreed to try to fix them.

It seems that `lispPackages.cl-fuse` may not be fixed before the merge to master. /cc @7c6f434c

Ericson2314 commented:
Can't fix this instant, but the crux of the Perl packages' problem is that `ld` no longer magically see's into the C(XX)'s compiler's lib directory. `LD=$CC` might be the best solution to avoid violating layers.

orivej commented:
Thanks! I'm fixing perl packages in #32830.

7c6f434c commented:
Hm. I don't really understand where `NIX_LDFLAGS` gets lost in CL-Fuse build… So I gave up and the shared library is now prebuilt separately.
